src/Altinn.App.Core/Infrastructure/Clients/Storage/AuthenticationTokenResolver.cs

```diff
@@ -0,0 +1,114 @@
+using Altinn.App.Core.Configuration;
+using Altinn.App.Core.Features.Maskinporten;
+using Altinn.App.Core.Helpers;
+using Altinn.App.Core.Internal;
+using Altinn.App.Core.Internal.App;
+using Altinn.App.Core.Internal.Auth;
+using Altinn.App.Core.Models;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.App.Core.Infrastructure.Clients.Storage;
+
+/// <summary>
+/// Resolves authentication tokens based on the specified authentication method.
+/// </summary>
+public interface IAuthenticationTokenResolver
+{
+    /// <summary>
+    /// Retrieves an access token based on the specified authentication method.
+    /// </summary>
+    Task<JwtToken> GetAccessToken(
+        AuthenticationMethod authenticationMethod,
+        CancellationToken cancellationToken = default
+    );
+}
+
+internal class AuthenticationTokenResolver : IAuthenticationTokenResolver
+{
+    private readonly IUserTokenProvider _userTokenProvider;
+    private readonly IMaskinportenClient _maskinportenClient;
+    private readonly IAppMetadata _appMetadata;
+    private readonly IHttpClientFactory _httpClientFactory;
+    private readonly IOptionsMonitor<PlatformSettings> _platformSettings;
+
+    private readonly bool _isDev;
+
+    public AuthenticationTokenResolver(
+        IHttpClientFactory httpClientFactory,
+        IUserTokenProvider userTokenProvider,
+        IMaskinportenClient maskinportenClient,
+        IAppMetadata appMetadata,
+        IHostEnvironment hostEnvironment,
+        IOptionsMonitor<PlatformSettings> platformSettings
+    )
+    {
+        _userTokenProvider = userTokenProvider;
+        _maskinportenClient = maskinportenClient;
+        _appMetadata = appMetadata;
+        _httpClientFactory = httpClientFactory;
+        _platformSettings = platformSettings;
+        _isDev = hostEnvironment.IsDevelopment();
```

Comment:
I think we use a different heuristic for the `LocaltestValidation` thing, we should probably do the same thing and potentially pull it out into some common internal thing